Connect to DuckDB

In [1]:
# Load the extension
%load_ext sql

There's a new jupysql version available (0.10.14), you're running 0.10.12. To upgrade: pip install jupysql --upgrade
Deploy Dash apps for free on Ploomber Cloud! Learn more: https://ploomber.io/s/signup


In [2]:
# Connect to DuckDB
%sql duckdb:///tpch.db

Connecting to 'duckdb:///tpch.db'

In [3]:
%config SqlMagic.displaylimit = 100

In [4]:
%%sql
-- Run a simple show tables
SELECT
  table_name
FROM
  information_schema.tables
WHERE
  table_schema = 'main'

Running query in 'duckdb:///tpch.db'

+------------+
| table_name |
+------------+
|  customer  |
|  lineitem  |
|   nation   |
|   orders   |
|    part    |
|  partsupp  |
|   region   |
|  supplier  |
+------------+

In [ ]:
# If you do not see any tables run the below command (after uncommeting it)
#! python setup.py

# [GET DATA] Use SELECT...FROM, LIMIT, WHERE, & ORDER BY to read the required data from tables

The most common use for querying is to read data in our tables. We can do this using a `SELECT ... FROM` statement, as shown below.



In [ ]:
%%sql
SELECT
  *
FROM
  customer
LIMIT
  1;

In [ ]:
%%sql
-- use * to specify all columns
SELECT
  *
FROM
  orders
LIMIT
  5;

However, running a `SELECT ... FROM` statement can cause issues when the data set is extensive.


In [ ]:
%%sql
-- use * to specify all columns
SELECT
  *
FROM
  orders
LIMIT
  5;

In [ ]:
%%sql
-- use column c_names only to read data from those columns
SELECT
  o_orderkey,
  o_totalprice
FROM
  orders
LIMIT
  5;


We can use the `WHERE` clause if we want to get the rows that match specific criteria. We can specify one or more filters within the `WHERE` clause. 

The `WHERE` clause with more than one filter can use combinations of `AND` and `OR` criteria to combine the filter criteria, as shown below.



In [ ]:
%%sql
-- all customer rows that have c_nationkey = 20
SELECT
  *
FROM
  customer
WHERE
  c_nationkey = 20
LIMIT
  10;

In [ ]:
%%sql
-- all customer rows that have c_nationkey = 20 and c_acctbal > 1000
SELECT
  *
FROM
  customer
WHERE
  c_nationkey = 20
  AND c_acctbal > 1000
LIMIT
  10;

In [ ]:
%%sql
-- all customer rows that have c_nationkey = 20 or c_acctbal > 1000
SELECT
  *
FROM
  customer
WHERE
  c_nationkey = 20
  OR c_acctbal > 1000
LIMIT
  10;

In [ ]:
%%sql
-- all customer rows that have (c_nationkey = 20 and c_acctbal > 1000) or rows that have c_nationkey = 11
SELECT
  *
FROM
  customer
WHERE
  (
    c_nationkey = 20
    AND c_acctbal > 1000
  )
  OR c_nationkey = 11
LIMIT
  10;


We can combine multiple filter clauses, as seen above. We have seen examples of equals (`=`) and greater than (`>`) conditional operators. There are 6 **conditional operators**, they are

1. **<** Less than
2. **>** Greater than
3. **<=** Less than or equal to
4. **>=** Greater than or equal to
5. **=** Equal
6. **<>** and **!=** both represent Not equal (some DBs only support one of these)

Additionally, for string types, we can make **pattern matching with `like` condition**. In a `like` condition, a `_` means any single character, and `%` means zero or more characters, for example.




In [ ]:
%%sql
-- all customer rows where the c_name has a 381 in it
SELECT
  *
FROM
  customer
WHERE
  c_name LIKE '%381%';

In [ ]:
%%sql
-- all customer rows where the c_name ends with a 381
SELECT
  *
FROM
  customer
WHERE
  c_name LIKE '%381';

In [ ]:
%%sql
-- all customer rows where the c_name starts with a 381
SELECT
  *
FROM
  customer
WHERE
  c_name LIKE '381%';

In [ ]:
%%sql
-- all customer rows where the c_name has a combination of any character and 9 and 1
SELECT
  *
FROM
  customer
WHERE
  c_name LIKE '%_91%';

We can also filter for more than one value using `IN` and `NOT IN`.


In [ ]:
%%sql
-- all customer rows which have c_nationkey = 10 or c_nationkey = 20
SELECT
  *
FROM
  customer
WHERE
  c_nationkey IN (10, 20);

In [ ]:
%%sql
-- all customer rows which have do not have c_nationkey as 10 or 20
SELECT
  *
FROM
  customer
WHERE
  c_nationkey NOT IN (10, 20);

We can get the number of rows in a table using `count(*)` as shown below.


In [ ]:
%%sql
SELECT
  COUNT(*)
FROM
  customer;

In [ ]:
%%sql
SELECT
  COUNT(*)
FROM
  lineitem;

If we want to get the rows sorted by values in a specific column, we use `ORDER BY`, for example.


In [ ]:
%%sql
-- Will show the first ten customer records with the lowest custkey
-- rows are ordered in ASC order by default
SELECT
  *
FROM
  orders
ORDER BY
  o_custkey
LIMIT
  10;

In [ ]:
%%sql
-- Will show the first ten customer's records with the highest custkey
SELECT
  *
FROM
  orders
ORDER BY
  o_custkey DESC
LIMIT
  10;


# [JOINS] Combine data from multiple tables using JOINs (there are different types of JOINs)

We can combine data from multiple tables using joins. When we write a join query, we have a format as shown below.

```sql
-- not based on real tables
SELECT
    a.*
FROM
    table_a a -- LEFT table a
    JOIN table_b b -- RIGHT table b
    ON a.id = b.id
```

The table specified first (table_a) is the left table, whereas the table established second is the right table. When we have multiple tables joined, we consider the joined dataset from the first two tables as the left table and the third table as the right table (The DB optimizes the joins for performance).

```sql
-- not based on real tables
SELECT
    a.*
FROM
    table_a a -- LEFT table a
    JOIN table_b b -- RIGHT table b
    ON a.id = b.id
    JOIN table_c c -- LEFT table is the joined data from 
-- table_a & table_b, right table is table_c
    ON a.c_id = c.id
```

There are five main types of joins, they are:

## 1. Inner join (default): Get only rows in both tables



In [ ]:
%%sql
SELECT
  o.o_orderkey,
  l.l_orderkey
FROM
  orders o
  JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY
LIMIT
  100;

In [5]:
%%sql
SELECT
  COUNT(o.o_orderkey) AS order_rows_count,
  COUNT(l.l_orderkey) AS lineitem_rows_count
FROM
  orders o
  JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY;

Running query in 'duckdb:///tpch.db'

+------------------+---------------------+
| order_rows_count | lineitem_rows_count |
+------------------+---------------------+
|      247650      |        247650       |
+------------------+---------------------+

**Note:** JOIN defaults to INNER JOIN.

The output will have rows from orders and lineitem that found at least one matching row from the other table with the specified join condition (same o_orderkey and orderdate within ship date +/- 5 days). 

We can also see that 247650 rows from orders and lineitem tables matched.

## 2. Left outer join (aka left join): Get all rows from the left table and only matching rows from the right table.



In [6]:
%%sql
SELECT
  o.o_orderkey,
  l.l_orderkey
FROM
  orders o
  LEFT JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY
LIMIT
  100;

Running query in 'duckdb:///tpch.db'

+------------+------------+
| o_orderkey | l_orderkey |
+------------+------------+
|     7      |     7      |
|     32     |     32     |
|     33     |     33     |
|     69     |     69     |
|     71     |     71     |
|    133     |    133     |
|    198     |    198     |
|    259     |    259     |
|    260     |    260     |
|    288     |    288     |
|    353     |    353     |
|    354     |    354     |
|    420     |    420     |
|    482     |    482     |
|    485     |    485     |
|    486     |    486     |
|    512     |    512     |
|    517     |    517     |
|    548     |    548     |
|    576     |    576     |
|    576     |    576     |
|    581     |    581     |
|    614     |    614     |
|    673     |    673     |
|    678     |    678     |
|    679     |    679     |
|    737     |    737     |
|    771     |    771     |
|    771     |    771     |
|    772     |    772     |
|    774     |    774     |
|    899     |    899     |
|    930     |    930     |
|    930     |    930     |
|    962     |    962     |
|    962     |    962     |
|    992     |    992     |
|    992     |    992     |
|    1027    |    1027    |
|    1088    |    1088    |
|    1095    |    1095    |
|    1095    |    1095    |
|    1120    |    1120    |
|    1186    |    1186    |
|    1252    |    1252    |
|    1284    |    1284    |
|    1287    |    1287    |
|    1314    |    1314    |
|    1410    |    1410    |
|    1571    |    1571    |
|    1573    |    1573    |
|    1574    |    1574    |
|    1605    |    1605    |
|    1639    |    1639    |
|    1639    |    1639    |
|    1664    |    1664    |
|    1667    |    1667    |
|    1667    |    1667    |
|    1671    |    1671    |
|    1735    |    1735    |
|    1767    |    1767    |
|    1824    |    1824    |
|    1856    |    1856    |
|    1891    |    1891    |
|    1923    |    1923    |
|    1954    |    1954    |
|    1956    |    1956    |
|    1991    |    1991    |
|   122375   |   122375   |
|   122404   |   122404   |
|   122406   |   122406   |
|   122432   |   122432   |
|   122464   |   122464   |
|   122468   |   122468   |
|   122535   |   122535   |
|   122566   |   122566   |
|   122595   |   122595   |
|   122658   |   122658   |
|   122663   |   122663   |
|   122691   |   122691   |
|   122784   |   122784   |
|   122791   |   122791   |
|   122885   |   122885   |
|   122948   |   122948   |
|   122977   |   122977   |
|   122980   |   122980   |
|   123011   |   123011   |
|   123040   |   123040   |
|   123139   |   123139   |
|   123139   |   123139   |
|   123168   |   123168   |
|   123204   |   123204   |
|   123206   |   123206   |
|   123207   |   123207   |
|   123239   |   123239   |
|   123271   |   123271   |
|   123271   |   123271   |
|   123296   |   123296   |
|   123361   |   123361   |
|   123363   |   123363   |
+------------+------------+
Truncated to displaylimit of 100.

In [7]:
%%sql
SELECT
  COUNT(o.o_orderkey) AS order_rows_count,
  COUNT(l.l_orderkey) AS lineitem_rows_count
FROM
  orders o
  LEFT JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY;

Running query in 'duckdb:///tpch.db'

+------------------+---------------------+
| order_rows_count | lineitem_rows_count |
+------------------+---------------------+
|     1519332      |        247650       |
+------------------+---------------------+

The output will have all the rows from orders and the rows from lineitem that were able to find at least one matching row from the orders table with the specified join condition (same o_orderkey and orderdate within ship date +/- 5 days). 

We can also see that the number of rows from the orders table is 1,519,332 & from the lineitem table is 247,650. The number of rows in orders is 1,500,000, but the join condition produces 1,519,332 since some orders match with multiple lineitems.

## 3. Right outer join (aka right join): Get matching rows from the left and all rows from the right table.



In [8]:
%%sql
SELECT
  o.o_orderkey,
  l.l_orderkey
FROM
  orders o
  RIGHT JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY
LIMIT
  100;

Running query in 'duckdb:///tpch.db'

+------------+------------+
| o_orderkey | l_orderkey |
+------------+------------+
|     7      |     7      |
|     32     |     32     |
|     33     |     33     |
|     69     |     69     |
|     71     |     71     |
|    133     |    133     |
|    198     |    198     |
|    259     |    259     |
|    260     |    260     |
|    288     |    288     |
|    353     |    353     |
|    354     |    354     |
|    420     |    420     |
|    482     |    482     |
|    485     |    485     |
|    486     |    486     |
|    512     |    512     |
|    517     |    517     |
|    548     |    548     |
|    576     |    576     |
|    576     |    576     |
|    581     |    581     |
|    614     |    614     |
|    678     |    678     |
|    679     |    679     |
|    737     |    737     |
|    771     |    771     |
|    771     |    771     |
|    772     |    772     |
|    774     |    774     |
|    899     |    899     |
|    962     |    962     |
|    962     |    962     |
|    992     |    992     |
|    992     |    992     |
|    1027    |    1027    |
|    1088    |    1088    |
|    1094    |    1094    |
|    1095    |    1095    |
|    1095    |    1095    |
|    1120    |    1120    |
|    1186    |    1186    |
|    1252    |    1252    |
|    1284    |    1284    |
|    1314    |    1314    |
|    1410    |    1410    |
|    1506    |    1506    |
|    1506    |    1506    |
|    1571    |    1571    |
|    1573    |    1573    |
|    1574    |    1574    |
|    1600    |    1600    |
|    1605    |    1605    |
|    1639    |    1639    |
|    1639    |    1639    |
|    1664    |    1664    |
|    1667    |    1667    |
|    1667    |    1667    |
|    1671    |    1671    |
|    1735    |    1735    |
|    1767    |    1767    |
|    1794    |    1794    |
|    1824    |    1824    |
|    1825    |    1825    |
|    1856    |    1856    |
|    1891    |    1891    |
|    1923    |    1923    |
|    1954    |    1954    |
|    1956    |    1956    |
|    1991    |    1991    |
|   122375   |   122375   |
|   122404   |   122404   |
|   122406   |   122406   |
|   122432   |   122432   |
|   122464   |   122464   |
|   122468   |   122468   |
|   122535   |   122535   |
|   122566   |   122566   |
|   122595   |   122595   |
|   122658   |   122658   |
|   122663   |   122663   |
|   122784   |   122784   |
|   122791   |   122791   |
|   122885   |   122885   |
|   122948   |   122948   |
|   122977   |   122977   |
|   123011   |   123011   |
|   123040   |   123040   |
|   123139   |   123139   |
|   123139   |   123139   |
|   123168   |   123168   |
|   123204   |   123204   |
|   123206   |   123206   |
|   123207   |   123207   |
|   123239   |   123239   |
|   123271   |   123271   |
|   123271   |   123271   |
|   123296   |   123296   |
|   123361   |   123361   |
|   123363   |   123363   |
+------------+------------+
Truncated to displaylimit of 100.

In [17]:
%%sql
SELECT
  COUNT(o.o_orderkey) AS order_rows_count,
  COUNT(l.l_orderkey) AS lineitem_rows_count
FROM
  orders o
  RIGHT JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY;

Running query in 'duckdb:///tpch.db'

+------------------+---------------------+
| order_rows_count | lineitem_rows_count |
+------------------+---------------------+
|      247650      |       6001215       |
+------------------+---------------------+

The output will have the rows from orders that found at least one matching row from the lineitem table with the specified join condition (same o_orderkey and orderdate within ship date +/- 5 days) and all the rows from the lineitem table.

We can also see that the number of rows from the orders table is 247,650 & from the lineitem table is 6,001,215 .

## 4. Full outer join: Get all rows from both the left and right tables.



In [18]:
%%sql
SELECT
  o.o_orderkey,
  l.l_orderkey
FROM
  orders o
  FULL OUTER JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY
LIMIT
  100;

Running query in 'duckdb:///tpch.db'

+------------+------------+
| o_orderkey | l_orderkey |
+------------+------------+
|     7      |     7      |
|     32     |     32     |
|     33     |     33     |
|     69     |     69     |
|     71     |     71     |
|    133     |    133     |
|    198     |    198     |
|    259     |    259     |
|    260     |    260     |
|    288     |    288     |
|    353     |    353     |
|    354     |    354     |
|    420     |    420     |
|    482     |    482     |
|    485     |    485     |
|    486     |    486     |
|    512     |    512     |
|    517     |    517     |
|    548     |    548     |
|    576     |    576     |
|    576     |    576     |
|    581     |    581     |
|    614     |    614     |
|    673     |    673     |
|    678     |    678     |
|    679     |    679     |
|    737     |    737     |
|    771     |    771     |
|    771     |    771     |
|    772     |    772     |
|    774     |    774     |
|    899     |    899     |
|    962     |    962     |
|    962     |    962     |
|    992     |    992     |
|    992     |    992     |
|    1027    |    1027    |
|    1088    |    1088    |
|    1095    |    1095    |
|    1095    |    1095    |
|    1120    |    1120    |
|    1186    |    1186    |
|    1252    |    1252    |
|    1284    |    1284    |
|    1287    |    1287    |
|    1314    |    1314    |
|    1410    |    1410    |
|    1506    |    1506    |
|    1506    |    1506    |
|    1571    |    1571    |
|    1573    |    1573    |
|    1574    |    1574    |
|    1605    |    1605    |
|    1639    |    1639    |
|    1639    |    1639    |
|    1664    |    1664    |
|    1667    |    1667    |
|    1667    |    1667    |
|    1671    |    1671    |
|    1735    |    1735    |
|    1767    |    1767    |
|    1794    |    1794    |
|    1824    |    1824    |
|    1825    |    1825    |
|    1856    |    1856    |
|    1891    |    1891    |
|    1923    |    1923    |
|    1954    |    1954    |
|    1956    |    1956    |
|    1991    |    1991    |
|   122375   |   122375   |
|   122404   |   122404   |
|   122406   |   122406   |
|   122432   |   122432   |
|   122464   |   122464   |
|   122468   |   122468   |
|   122535   |   122535   |
|   122566   |   122566   |
|   122595   |   122595   |
|   122658   |   122658   |
|   122663   |   122663   |
|   122784   |   122784   |
|   122791   |   122791   |
|   122885   |   122885   |
|   122948   |   122948   |
|   122977   |   122977   |
|   122980   |   122980   |
|   123011   |   123011   |
|   123040   |   123040   |
|   123139   |   123139   |
|   123139   |   123139   |
|   123168   |   123168   |
|   123204   |   123204   |
|   123206   |   123206   |
|   123207   |   123207   |
|   123239   |   123239   |
|   123271   |   123271   |
|   123271   |   123271   |
|   123296   |   123296   |
|   123361   |   123361   |
+------------+------------+
Truncated to displaylimit of 100.

In [19]:
%%sql
SELECT
  COUNT(o.o_orderkey) AS order_rows_count,
  COUNT(l.l_orderkey) AS lineitem_rows_count
FROM
  orders o
  FULL OUTER JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY;

Running query in 'duckdb:///tpch.db'

+------------------+---------------------+
| order_rows_count | lineitem_rows_count |
+------------------+---------------------+
|     1519332      |       6001215       |
+------------------+---------------------+

The output will have all the rows from orders that found at least one matching row from the lineitem table with the specified join condition (same o_orderkey and orderdate within ship date +/- 5 days) and all the rows from the lineitem table.

We can also see that the number of rows from the orders table is 1,519,332 & from the lineitem table is 6,001,215.

## 5. Cross join: Get the cartesian product of all rows



In [20]:
%%sql
SELECT
  n.n_name AS nation_c_name,
  r.r_name AS region_c_name
FROM
  nation n
  CROSS JOIN region r;

Running query in 'duckdb:///tpch.db'

+----------------+---------------+
| nation_c_name  | region_c_name |
+----------------+---------------+
|    ALGERIA     |     AFRICA    |
|   ARGENTINA    |     AFRICA    |
|     BRAZIL     |     AFRICA    |
|     CANADA     |     AFRICA    |
|     EGYPT      |     AFRICA    |
|    ETHIOPIA    |     AFRICA    |
|     FRANCE     |     AFRICA    |
|    GERMANY     |     AFRICA    |
|     INDIA      |     AFRICA    |
|   INDONESIA    |     AFRICA    |
|      IRAN      |     AFRICA    |
|      IRAQ      |     AFRICA    |
|     JAPAN      |     AFRICA    |
|     JORDAN     |     AFRICA    |
|     KENYA      |     AFRICA    |
|    MOROCCO     |     AFRICA    |
|   MOZAMBIQUE   |     AFRICA    |
|      PERU      |     AFRICA    |
|     CHINA      |     AFRICA    |
|    ROMANIA     |     AFRICA    |
|  SAUDI ARABIA  |     AFRICA    |
|    VIETNAM     |     AFRICA    |
|     RUSSIA     |     AFRICA    |
| UNITED KINGDOM |     AFRICA    |
| UNITED STATES  |     AFRICA    |
|    ALGERIA     |    AMERICA    |
|   ARGENTINA    |    AMERICA    |
|     BRAZIL     |    AMERICA    |
|     CANADA     |    AMERICA    |
|     EGYPT      |    AMERICA    |
|    ETHIOPIA    |    AMERICA    |
|     FRANCE     |    AMERICA    |
|    GERMANY     |    AMERICA    |
|     INDIA      |    AMERICA    |
|   INDONESIA    |    AMERICA    |
|      IRAN      |    AMERICA    |
|      IRAQ      |    AMERICA    |
|     JAPAN      |    AMERICA    |
|     JORDAN     |    AMERICA    |
|     KENYA      |    AMERICA    |
|    MOROCCO     |    AMERICA    |
|   MOZAMBIQUE   |    AMERICA    |
|      PERU      |    AMERICA    |
|     CHINA      |    AMERICA    |
|    ROMANIA     |    AMERICA    |
|  SAUDI ARABIA  |    AMERICA    |
|    VIETNAM     |    AMERICA    |
|     RUSSIA     |    AMERICA    |
| UNITED KINGDOM |    AMERICA    |
| UNITED STATES  |    AMERICA    |
|    ALGERIA     |      ASIA     |
|   ARGENTINA    |      ASIA     |
|     BRAZIL     |      ASIA     |
|     CANADA     |      ASIA     |
|     EGYPT      |      ASIA     |
|    ETHIOPIA    |      ASIA     |
|     FRANCE     |      ASIA     |
|    GERMANY     |      ASIA     |
|     INDIA      |      ASIA     |
|   INDONESIA    |      ASIA     |
|      IRAN      |      ASIA     |
|      IRAQ      |      ASIA     |
|     JAPAN      |      ASIA     |
|     JORDAN     |      ASIA     |
|     KENYA      |      ASIA     |
|    MOROCCO     |      ASIA     |
|   MOZAMBIQUE   |      ASIA     |
|      PERU      |      ASIA     |
|     CHINA      |      ASIA     |
|    ROMANIA     |      ASIA     |
|  SAUDI ARABIA  |      ASIA     |
|    VIETNAM     |      ASIA     |
|     RUSSIA     |      ASIA     |
| UNITED KINGDOM |      ASIA     |
| UNITED STATES  |      ASIA     |
|    ALGERIA     |     EUROPE    |
|   ARGENTINA    |     EUROPE    |
|     BRAZIL     |     EUROPE    |
|     CANADA     |     EUROPE    |
|     EGYPT      |     EUROPE    |
|    ETHIOPIA    |     EUROPE    |
|     FRANCE     |     EUROPE    |
|    GERMANY     |     EUROPE    |
|     INDIA      |     EUROPE    |
|   INDONESIA    |     EUROPE    |
|      IRAN      |     EUROPE    |
|      IRAQ      |     EUROPE    |
|     JAPAN      |     EUROPE    |
|     JORDAN     |     EUROPE    |
|     KENYA      |     EUROPE    |
|    MOROCCO     |     EUROPE    |
|   MOZAMBIQUE   |     EUROPE    |
|      PERU      |     EUROPE    |
|     CHINA      |     EUROPE    |
|    ROMANIA     |     EUROPE    |
|  SAUDI ARABIA  |     EUROPE    |
|    VIETNAM     |     EUROPE    |
|     RUSSIA     |     EUROPE    |
| UNITED KINGDOM |     EUROPE    |
| UNITED STATES  |     EUROPE    |
+----------------+---------------+
Truncated to displaylimit of 100.


The output will have every row of the nation joined with every row of the region. There are 25 nations and five regions, leading to 125 rows in our result from the cross-join.

![Joins](./images/im_27_1.png)
![Cross join](./images/im_27_2.png)

There are cases where we will need to join a table with itself, called a SELF-join.

**Example**:

1. For every customer order, get the order placed earlier in the same week (Sunday - Saturday, not the previous seven days). Only show customer orders that have at least one such order.



In [21]:
%%sql
SELECT
  o1.o_custkey
FROM
  orders o1
  JOIN orders o2 ON o1.o_custkey = o2.o_custkey
  AND YEAR (o1.o_orderdate) = YEAR (o2.o_orderdate)
  AND week (o1.o_orderdate) = week (o2.o_orderdate)
WHERE
  o1.o_orderkey != o2.o_orderkey;

Running query in 'duckdb:///tpch.db'

+-----------+
| o_custkey |
+-----------+
|   130057  |
|    9220   |
|   93385   |
|   123823  |
|   55126   |
|   39907   |
|   50417   |
|   11704   |
|   64754   |
|   38299   |
|   134617  |
|   31696   |
|    3478   |
|   116326  |
|   108556  |
|   105941  |
|   28924   |
|   133400  |
|    4306   |
|   65312   |
|   29749   |
|   38071   |
|   44923   |
|   99505   |
|   36427   |
|   71026   |
|   18637   |
|   54376   |
|   61123   |
|   73696   |
|   77266   |
|    5663   |
|   48157   |
|    4243   |
|   92359   |
|   111901  |
|   33227   |
|   88393   |
|   101821  |
|   77359   |
|   87323   |
|   20941   |
|   31168   |
|   104617  |
|   103651  |
|   24016   |
|   121615  |
|   66976   |
|   60367   |
|   13177   |
|   98308   |
|   38041   |
|   27061   |
|   80107   |
|   63397   |
|   121156  |
|   68575   |
|   17677   |
|   138236  |
|   44716   |
|   100228  |
|   22951   |
|   46567   |
|   89615   |
|    7828   |
|   147937  |
|   98675   |
|   148414  |
|   93235   |
|   53362   |
|   15377   |
|   56425   |
|   110447  |
|   83902   |
|   26557   |
|   33688   |
|   130417  |
|   123316  |
|   108506  |
|   101696  |
|   23443   |
|   22462   |
|   41701   |
|   53116   |
|   149798  |
|   58663   |
|   45688   |
|   138322  |
|   121237  |
|   81229   |
|   112495  |
|   98765   |
|   78286   |
|   53495   |
|   15310   |
|   59477   |
|   148771  |
|   78955   |
|   146293  |
|   116053  |
+-----------+
Truncated to displaylimit of 100.


Most analytical queries require calculating metrics that involve combining data from multiple rows. `GROUP BY` allows us to perform aggregate calculations on data from a set of rows grouped together based on values of specified column(s).

**Example**:

1. Create a report that shows the number of orders per orderpriority segment.



In [22]:
%%sql
SELECT
  o_orderpriority,
  COUNT(*) AS num_orders
FROM
  orders
GROUP BY
  o_orderpriority;

Running query in 'duckdb:///tpch.db'

+-----------------+------------+
| o_orderpriority | num_orders |
+-----------------+------------+
| 4-NOT SPECIFIED |   300254   |
|      5-LOW      |   300589   |
|     1-URGENT    |   300343   |
|     3-MEDIUM    |   298723   |
|      2-HIGH     |   300091   |
+-----------------+------------+


In the above query, we group the data by `orderpriority`, and the calculation `count(*)` will be applied to the rows having a specific `orderpriority` value. The output will consist of one row per unique value of `orderpriority` and the `count(*)` calculation.

![Group by](./images/im_28.png)

The calculations allowed are typically SUM/MIN/MAX/AVG/COUNT. However, some databases have more complex aggregate functions; check your DB documentation.

# [SUB QUERY] Use the result of a query within a query using sub-queries

When we want to use the result of a query as a table in another query, we use subqueries. **Example**:

1. Create a report that shows the nation, how many items it supplied (by suppliers in that nation), and how many items it purchased (by customers in that nation). 



In [23]:
%%sql
SELECT
  n.n_name AS nation_c_name,
  s.quantity AS supplied_items_quantity,
  c.quantity AS purchased_items_quantity
FROM
  nation n
  LEFT JOIN (
    SELECT
      n.n_nationkey,
      SUM(l.l_quantity) AS quantity
    FROM
      lineitem l
      JOIN supplier s ON l.l_suppkey = s.s_suppkey
      JOIN nation n ON s.s_nationkey = n.n_nationkey
    GROUP BY
      n.n_nationkey
  ) s ON n.n_nationkey = s.n_nationkey
  LEFT JOIN (
    SELECT
      n.n_nationkey,
      SUM(l.l_quantity) AS quantity
    FROM
      lineitem l
      JOIN orders o ON l.l_orderkey = o.o_orderkey
      JOIN customer c ON o.o_custkey = c.c_custkey
      JOIN nation n ON c.c_nationkey = n.n_nationkey
    GROUP BY
      n.n_nationkey
  ) c ON n.n_nationkey = c.n_nationkey;

Running query in 'duckdb:///tpch.db'

+----------------+-------------------------+--------------------------+
| nation_c_name  | supplied_items_quantity | purchased_items_quantity |
+----------------+-------------------------+--------------------------+
|    ALGERIA     |        6454691.00       |        6117618.00        |
|   ARGENTINA    |        6339724.00       |        6087566.00        |
|     BRAZIL     |        6085551.00       |        6149174.00        |
|     CANADA     |        6296547.00       |        6168913.00        |
|     EGYPT      |        6385468.00       |        6024134.00        |
|    ETHIOPIA    |        5817697.00       |        6095241.00        |
|     FRANCE     |        6141618.00       |        6289987.00        |
|    GERMANY     |        6076474.00       |        6098776.00        |
|     INDIA      |        6347392.00       |        6102406.00        |
|   INDONESIA    |        6204759.00       |        6276420.00        |
|      IRAN      |        5992033.00       |        6091271.00        |
|      IRAQ      |        6678487.00       |        6003025.00        |
|     JAPAN      |        5775308.00       |        6067924.00        |
|     JORDAN     |        5539782.00       |        6230581.00        |
|     KENYA      |        5753471.00       |        6002935.00        |
|    MOROCCO     |        5691519.00       |        6060121.00        |
|   MOZAMBIQUE   |        6253676.00       |        6243695.00        |
|      PERU      |        6412494.00       |        6035782.00        |
|     CHINA      |        6205793.00       |        6179322.00        |
|    ROMANIA     |        6079467.00       |        6212841.00        |
|  SAUDI ARABIA  |        6299506.00       |        5948921.00        |
|    VIETNAM     |        6111369.00       |        6149667.00        |
|     RUSSIA     |        6152008.00       |        6265778.00        |
| UNITED KINGDOM |        5946859.00       |        6046145.00        |
| UNITED STATES  |        6037102.00       |        6130552.00        |
+----------------+-------------------------+--------------------------+


In the above query, we can see that there are two sub-queries, one to calculate the quantity supplied by a nation and the other to calculate the quantity purchased by the customers of a nation.

# [CASE WHEN] Replicate IF.ELSE logic with CASE statements

We can do conditional logic in the `SELECT ... FROM` part of our query, as shown below.



In [24]:
%%sql
SELECT
    o_orderkey,
    o_totalprice,
    CASE
        WHEN o_totalprice > 100000 THEN 'high'
        WHEN o_totalprice BETWEEN 25000
        AND 100000 THEN 'medium'
        ELSE 'low'
    END AS order_price_bucket
FROM
    orders;

Running query in 'duckdb:///tpch.db'

+------------+--------------+--------------------+
| o_orderkey | o_totalprice | order_price_bucket |
+------------+--------------+--------------------+
|     1      |  173665.47   |        high        |
|     2      |   46929.18   |       medium       |
|     3      |  193846.25   |        high        |
|     4      |   32151.78   |       medium       |
|     5      |  144659.20   |        high        |
|     6      |   58749.59   |       medium       |
|     7      |  252004.18   |        high        |
|     32     |  208660.75   |        high        |
|     33     |  163243.98   |        high        |
|     34     |   58949.67   |       medium       |
|     35     |  253724.56   |        high        |
|     36     |   68289.96   |       medium       |
|     37     |  206680.66   |        high        |
|     38     |   82500.05   |       medium       |
|     39     |  341734.47   |        high        |
|     64     |   39414.99   |       medium       |
|     65     |  110643.60   |        high        |
|     66     |  103740.67   |        high        |
|     67     |  169405.01   |        high        |
|     68     |  330793.52   |        high        |
|     69     |  197689.49   |        high        |
|     70     |  113534.42   |        high        |
|     71     |  276992.74   |        high        |
|     96     |   68989.90   |       medium       |
|     97     |  110512.84   |        high        |
|     98     |   69168.33   |       medium       |
|     99     |  112126.95   |        high        |
|    100     |  187782.63   |        high        |
|    101     |  124906.11   |        high        |
|    102     |  164529.10   |        high        |
|    103     |  126990.79   |        high        |
|    128     |   66195.16   |       medium       |
|    129     |  261013.14   |        high        |
|    130     |  189484.12   |        high        |
|    131     |  130464.09   |        high        |
|    132     |  144947.21   |        high        |
|    133     |  114663.57   |        high        |
|    134     |  200354.30   |        high        |
|    135     |  213713.99   |        high        |
|    160     |  124132.06   |        high        |
|    161     |   34439.93   |       medium       |
|    162     |   2726.45    |        low         |
|    163     |  202379.28   |        high        |
|    164     |  301925.76   |        high        |
|    165     |  188023.67   |        high        |
|    166     |  154451.92   |        high        |
|    167     |   71124.35   |       medium       |
|    192     |  167668.42   |        high        |
|    193     |   88699.17   |       medium       |
|    194     |  176707.84   |        high        |
|    195     |  191542.31   |        high        |
|    196     |   47095.31   |       medium       |
|    197     |  163700.57   |        high        |
|    198     |  176189.31   |        high        |
|    199     |  112986.49   |        high        |
|    224     |  259960.36   |        high        |
|    225     |  220441.09   |        high        |
|    226     |  313134.62   |        high        |
|    227     |   54880.58   |       medium       |
|    228     |   2749.87    |        low         |
|    229     |  213076.69   |        high        |
|    230     |  171318.37   |        high        |
|    231     |  191808.17   |        high        |
|    256     |  130775.45   |        high        |
|    257     |   8649.81    |        low         |
|    258     |  294182.42   |        high        |
|    259     |  123298.34   |        high        |
|    260     |  272353.52   |        high        |
|    261     |  319306.86   |        high        |
|    262     |  143146.21   |        high        |
|    263     |  149369.68   |        high        |
|    288     |  266047.60   |        high        |
|    289     |  211162.52   |        high        |
|    290     |   99019.42   |       medium       |
|    291     |  108550.11   |        high        |
|    292     |   40637

We can see how we display different values depending on the `totalprice` column. We can also use multiple criteria as our conditional criteria (e.g., totalprice > 100000 AND orderpriority = '2-HIGH').

# [STANDARD FUNCTION] Use standard inbuilt DB functions for common String, Time, and Numeric data manipulation

When processing data, more often than not, we will need to change values in columns; shown below are a few standard functions to be aware of:

1. **String functions**
   1. **LENGTH** is used to calculate the length of a string. E.g., `SELECT LENGTH('hi');` will output 2.
   2. **CONCAT** combines multiple string columns into one. E.g., `SELECT CONCAT(o_orderstatus, '-', o_orderpriority) FROM ORDERS LIMIT 5;` will concatenate the o_orderstatus and o_orderpriority columns with a dash in between them.
   3. **SPLIT** is used to split a value into an array based on a given delimiter. E.g., `SELECT STRING_SPLIT(o_orderpriority, '-') FROM ORDERS LIMIT 5;` will output a column with arrays formed by splitting o_orderpriority values on `-`.
   4. **SUBSTRING** is used to get a sub-string from a value, given the start and end character indices. E.g., `SELECT o_orderpriority, SUBSTRING(o_orderpriority, 1, 5) FROM orders LIMIT 5;` will get the first five (1 - 5) characters of the o_orderpriority column.
   5. **TRIM** is used to remove empty spaces to the left and right of the value. E.g., `SELECT TRIM(' hi ');` will output `hi` without any spaces around it. LTRIM and RTRIM are similar but only remove spaces before and after the string, respectively.
2. **Date and Time functions**
   1. **Adding and subtracting dates**: Is used to add and subtract periods; the format heavily depends on the DB. In DuckDB, `date_diff` accepts 3 parameters, the outputs unit (day, month, year), the datetime/date values a and b such that the output will be a - b. The `+ INTERVAL n UNIT(DAY/MONTH/YEAR)` adds the value (in specified units) to the timestamp value.
      ```sql
        -- Date and Time Functions
        SELECT
            date_diff('day', DATE '2022-10-01', DATE '2023-11-05') AS diff_in_days,
            date_diff('month', DATE '2022-10-01', DATE '2023-11-05') AS diff_in_months,
            date_diff('year', DATE '2022-10-01', DATE '2023-11-05') AS diff_in_years,
            DATE '2022-10-01' + INTERVAL 400 DAY AS new_date;
      ```
    It will show the difference between the two dates in the specified period. We can also add/subtract an arbitrary period from a date/time column. E.g., `SELECT DATE '2022-11-05' + INTERVAL '10' DAY;` will show the output `2022-11-15` (try subtraction of dates).
   2. **String <=> date/time conversions**: When we want to change the data type of a string to date/time, we can use the `DATE 'YYYY-MM-DD'` or `TIMESTAMP 'YYYY-MM-DD HH:mm:SS'` functions. But when the data is in a non-standard date/time format such as `MM/DD/YYYY`, we will need to specify the input structure; we do this using `strptime`, E.g., `SELECT strptime('11-05-2023', '%m-%d-%Y');`.  
   
   We can convert a timestamp/date into a string with the required format using `strftime`. E.g., `SELECT strftime(o_orderdate, '%Y-%m-01') AS first_month_date FROM orders LIMIT 5;` will map every o_orderdate to the first of their month.

   See [this page](https://duckdb.org/docs/sql/functions/date) on how to set the proper date time format.
   3. **Time frame functions (YEAR/MONTH/DAY)**:  When we want to extract specific periods from a date/time column, we can use these functions. E.g., `SELECT year(DATE '2023-11-05');` will return 2023. Similarly, we have month, day, hour, min, etc.
3. **Numeric**
   1. **ROUND** is used to specify the number of digits allowed after the decimal point. E.g. `SELECT ROUND(100.102345, 2);`
   2. **ABS** is used to get the absolute value of a given number. E.g. `SELECT ABS(-100), ABS(100);` 
   3. **Mathematical operations** these are +,-,*,/.
   4. **Ceil/Floor** is used to get the next higher and most recent lower integers, given a decimal digit. E.g. `SELECT CEIL(100.1), FLOOR(100.1);`